feat: Add Reservation Create API

zerone1202 opened the pull request:
## #️⃣ Issue Number<!--- ex) #이슈번호, #이슈번호 -->

- Closes #87 
- 사용자가 게시글에 대해 예약을 생성할 수 있도록 하는 예약 요청 API를 구현

## 📝 요약(Summary)<!--- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
- DTO 추가
  - ReservationCreateRequest
  - ReservationCreateResponse

- Command Service 구현
  - ReservationCommandService
  - ReservationCommandServiceImpl

- Controller 구현
  - ReservationController
  - ReservationControllerImpl
  - POST /api/v1/reservations
  - 인증된 사용자(authUser.userId) 기준으로 예약 생성

## 🛠️ PR 유형
어떤 변경 사항이 있나요?
- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## ✅ PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.
- [x] 커밋 메시지 컨벤션에 맞게 작성

zerone1202 committed:
feat(reservation): Add ReservationCreateRequest DTO

- 예약 생성 요청을 위한 ReservationCreateRequest 추가
- postId, reservationStartAt, reservationEndAt, price, deposit 필드 포함
- 필수값 검증을 위한 @NotNull, @NotBlank 적용

zerone1202 committed:
feat(reservation): Add ReservationCreateResponse DTO

- 예약 생성 응답을 위한 ReservationCreateResponse 생성
- reservationId, postId, hostUserId, guestUserId, reservationStartAt, reservationEndAt 필드 포함

zerone1202 committed:
feat(reservation): Add createReservation() to ReservationCommandService

- 예약 생성 기능 정의를 위해 ReservationCommandService 인터페이스에 createReservation 메서드 추가
- 요청 DTO와 사용자 ID를 기반으로 예약 생성 책임을 명확히 분리

zerone1202 committed:
feat(reservation): Implement reservation creation logic in ReservationCommandServiceImpl

- 예약 생성 비즈니스 로직 구현
- PostQueryService, UserQueryService를 통한 게시글/사용자 조회
- Host/Guest 설정 및 ReservationStatus.REQUESTED 기본 상태 부여
- Reservation 엔티티 저장 후 ReservationCreateResponse로 변환하여 반환

zerone1202 committed:
feat(controller): Add reservation create API interface

- 예약 생성 API 시그니처 추가
- AuthUser 기반 인증 사용자 정보 전달 구조 반영

zerone1202 committed:
feat(controller): Implement reservation create API

- 예약 생성 API 구현
- 인증 사용자 정보(AuthUser) 기반 guestUser 매핑
- ReservationCommandService 호출하여 예약 생성 처리
- 생성된 예약 정보 ReservationCreateResponse 로 반환

zerone1202 committed:
feat(reservation): Add ReservationSuccessMessage for reservation creation

- 예약 생성 성공 메시지 추가
- 성공 응답 시 메시지 일관성을 유지하도록 ReservationSuccessMessage에 CREATE_SUCCESS 메시지 정의

zerone1202 committed:
Merge branch 'dev' of https://github.com/sparta-finale/golfrental into feat/87-reservation
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

zerone1202 committed:
Merge branch 'dev' of https://github.com/sparta-finale/golfrental into feat/87-reservation
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

zerone1202 commented on src/main/java/com/golfRental/domain/reservation/dto/request/ReservationCreateRequest.java:
```diff
@@ -0,0 +1,23 @@
+    private String reservationStartAt;
+
+    @NotNull(message = "예약 종료일은 필수입니다.")
+    private String reservationEndAt;
```
반영하겠습니다!

zerone1202 commented on src/main/java/com/golfRental/domain/reservation/service/command/ReservationCommandServiceImpl.java:
```diff
@@ -0,0 +1,64 @@
+        User hostUser = post.getUser();
+
+        LocalDateTime startAt = LocalDateTime.parse(request.getReservationStartAt());
+        LocalDateTime endAt = LocalDateTime.parse(request.getReservationEndAt());
```
추후 리팩토링 시 추가 예정입니다.
이슈 #108  참고 부탁드립니다!

zerone1202 commented on src/main/java/com/golfRental/domain/reservation/controller/ReservationControllerImpl.java:
```diff
@@ -2,36 +2,55 @@
+        ReservationCreateResponse response =
+                reservationCommandService.createReservation(request, authUser.getUserId());
+
+        return CommonApiResponse.created(
+                response,
+                ReservationSuccessMessage.RESERVATION_CREATED
+        );
```
추후 리팩토링시 전체적으로 반영 예정입니다!

moominIsCute commented on src/main/java/com/golfRental/domain/reservation/service/command/ReservationCommandServiceImpl.java:
예외 처리 시 에러코드를 사용하면 프론트엔드에서 구분하기 쉽습니다. 예외종류도 커스텀하신것이 있는것으로 알고 있는데 사용 안하신 이유가 있을까요?

moominIsCute replied:
다른리뷰어의 댓글에서 예외처리 관련하여 이슈넘버108에서 진행하신다는 것을 확인했습니다.

zerone1202 committed:
refactor(reservation): Update create request date fields to LocalDateTime

- ReservationCreateRequest의 reservationStartAt, reservationEndAt 타입을 String → LocalDateTime으로 변경
- Jackson이 날짜 파싱을 자동 처리하도록 개선
- ervice 레이어에서 불필요한 LocalDateTime.parse() 로직 제거
- 잘못된 날짜 형식 입력 시 400 Bad Request로 일관 처리되도록 개선

zerone1202 commented on src/main/java/com/golfRental/domain/reservation/dto/request/ReservationCreateRequest.java:
```diff
@@ -0,0 +1,23 @@
+    @NotNull(message = "예약 종료일은 필수입니다.")
+    private String reservationEndAt;
+
+    @NotNull(message = "가격은 필수입니다.")
+    private Integer price;
+
+    @NotNull(message = "보증금은 필수입니다.")
+    private Integer deposit;
```
Post 금액과 Reservation 금액의 의미가 다르다고 생각합니다!
Post.price / Post.deposit은 “게시글의 기본 가격/보증금”이고,
Reservation.price / Reservation.deposit은 “실제로 거래 시점에 확정된 금액”으로 서로 독립된 스펙이라 꼭 같은 타입일 필요는 없습니다.

또한, 할인, 부가세, 금액 연산 등이 없는 단순 저장 필드라
BigDecimal을 도입했을 때 실질적 이점이 크지 않습니다.

하지만, 일관성을 위해 Post 담당자와 논의 후 추후 리팩토링 진행하도록 하겠습니다.

zerone1202 commented on src/main/java/com/golfRental/domain/reservation/service/command/ReservationCommandServiceImpl.java:
```diff
@@ -0,0 +1,64 @@
+
+        Reservation saved = reservationRepository.save(reservation);
+
+        return ReservationCreateResponse.builder()
+                .reservationId(saved.getId())
+                .postId(post.getId())
+                .hostUserId(hostUser.getId())
+                .guestUserId(guestUser.getId())
+                .reservationStartAt(saved.getReservationStartAt())
+                .reservationEndAt(saved.getReservationEndAt())
+                .price(saved.getPrice())
+                .deposit(saved.getDeposit())
+                .status(saved.getStatus())
+                .build();
```
ReservationCreateResponse는 팀 내부적으로 DTO 생성에 Builder 패턴을 일관되게 사용하고 있습니다!

zerone1202 committed:
refactor(reservation): Update create request date fields to LocalDateTime

- ReservationCreateRequest의 reservationStartAt, reservationEndAt 타입을 String → LocalDateTime으로 변경
- Jackson이 날짜 파싱을 자동 처리하도록 개선
- ervice 레이어에서 불필요한 LocalDateTime.parse() 로직 제거
- 잘못된 날짜 형식 입력 시 400 Bad Request로 일관 처리되도록 개선

moominIsCute commented:
고생하셨습니다.
